src/utils/gasPriceManager.ts

```diff
@@ -0,0 +1,450 @@
+            this.logger.error(
+                "failed to fetch legacy gasPrices from estimateFeesPerGas",
+                { error: e }
+            )
+            gasPrice = undefined
+        }
+
+        if (gasPrice === undefined) {
+            this.logger.warn("gasPrice is undefined, using fallback value")
+            try {
+                gasPrice = await this.publicClient.getGasPrice()
+            } catch (e) {
+                this.logger.error("failed to get fallback gasPrice")
+                sentry.captureException(e)
+                throw e
+            }
+        }
+
+        return {
+            maxFeePerGas: gasPrice,
+            maxPriorityFeePerGas: gasPrice
+        }
+    }
+
+    private async estimateGasPrice(): Promise<GasPriceParameters> {
+        let maxFeePerGas: bigint | undefined
+        let maxPriorityFeePerGas: bigint | undefined
+        try {
+            const fees = await this.publicClient.estimateFeesPerGas({
+                chain: this.chain
+            })
+            maxFeePerGas = fees.maxFeePerGas
+            maxPriorityFeePerGas = fees.maxPriorityFeePerGas
+        } catch (e) {
+            sentry.captureException(e)
+            this.logger.error(
+                "failed to fetch eip-1559 gasPrices from estimateFeesPerGas",
+                { error: e }
+            )
+            maxFeePerGas = undefined
+            maxPriorityFeePerGas = undefined
+        }
+
+        if (maxPriorityFeePerGas === undefined) {
+            this.logger.warn(
+                "maxPriorityFeePerGas is undefined, using fallback value"
+            )
+            try {
+                maxPriorityFeePerGas =
+                    await this.getFallBackMaxPriorityFeePerGas(
+                        this.publicClient,
+                        maxFeePerGas ?? 0n
+                    )
+            } catch (e) {
+                this.logger.error("failed to get fallback maxPriorityFeePerGas")
+                sentry.captureException(e)
+                throw e
+            }
+        }
+
+        if (maxFeePerGas === undefined) {
+            this.logger.warn("maxFeePerGas is undefined, using fallback value")
+            try {
+                maxFeePerGas =
+                    (await this.getNextBaseFee(this.publicClient)) +
+                    maxPriorityFeePerGas
+            } catch (e) {
+                this.logger.error("failed to get fallback maxFeePerGas")
+                sentry.captureException(e)
+                throw e
+            }
+        }
+
+        if (maxPriorityFeePerGas === 0n) {
+            maxPriorityFeePerGas = maxFeePerGas / 200n
+        }
+
+        return { maxFeePerGas, maxPriorityFeePerGas }
+    }
+
+    private saveMaxFeePerGas(gasPrice: bigint, timestamp: number) {
+        const queue = this.queueMaxFeePerGas
+        const last = queue.length > 0 ? queue[queue.length - 1] : null
+
+        if (!last || timestamp - last.timestamp >= 1000) {
```

Comment:
wait if `!last` that means queue.length === 0, that's not supposed to be that way, right?

Reply:
This will be happen when we restart the server